pygsti/protocols/gst.py

```diff
@@ -1869,6 +1949,181 @@ def _add_badfit_estimates(results, base_estimate_label, badfit_options,
+    # Compute the various thresholds
+    ds_dof = ds.degrees_of_freedom(global_circuits_to_use)
+    nparams = model.num_modeltest_params  # just use total number of params
+    percentile = 0.025
+    nboxes = len(global_circuits_to_use)
+
+    two_dlogl_threshold = _chi2.ppf(1 - percentile, max(ds_dof - nparams, 1))
+    redbox_threshold  = _chi2.ppf(1 - percentile / nboxes, 1)
+    
+    # Not in cache, recompute
+    ddists = _compute_ddists(estimate)
+    
+    critical_percircuit_budgets = _opt.wildcardopt._get_critical_circuit_budgets(mdc_objfn, redbox_threshold)
+
+    #In Stefan's original code these were all passed in as a dictionary, instead pass these in as arguments.
+    
+    alpha = _bisect_alpha(0.1, 1e-3, mdc_objfn, two_dlogl_threshold,
+                          redbox_threshold, ddists, critical_percircuit_budgets, printer)
+   
+    return _ddist_wildcard_model(alpha, ddists)
+    
+def _bisect_alpha(guess, tol, mdc_objfn, two_dlogl_threshold, redbox_threshold, 
+                  ddists, critical_percircuit_budgets, printer):
+    
+    left = None
+    right = None
+    
+    while left is None or right is None:
+        printer.log(f'Searching for interval [{left}, {right}] with guess {guess}', 2)
+        # Test for feasibility
+        guessval = _test_feasible_alphas([guess], mdc_objfn, two_dlogl_threshold, 
+                                        redbox_threshold, ddists, critical_percircuit_budgets)
+        if guessval[0]:
+            printer.log('Guess value is feasible, ', 2)
+            left = guess
+            guess = left/2
+        else:
+            printer.log('Guess value is infeasible, ', 2)
+            right = guess
+            guess = 2*right
+    printer.log('Interval found!',2)
+    
+    # We now have an interval containing the crossover point
+    # Perform bisection
+    while abs(left - right) > tol:
+        printer.log(f'Performing bisection on interval [{left}, {right}]',2)
+        test = left - (left - right)/2.0
+        
+        testval = _test_feasible_alphas([test], mdc_objfn, two_dlogl_threshold, 
+                                        redbox_threshold, ddists, critical_percircuit_budgets)
+        
+        if testval[0]:
+            # Feasible, so shift left down
+            printer.log(f'Test value is feasible, ', 2)
+            left = test
+        else:
+            printer.log(f'Test value is infeasible, ', 2)
+            right = test
+            
+    printer.log('Interval within tolerance!',2)
+        
+    return left # Return the feasible one
+    
+def _test_feasible_alphas(alphas, mdc_objfn, two_dlogl_threshold, redbox_threshold, ddists, critical_percircuit_budgets):
+    
+    if not isinstance(alphas, list):
+        alphas = [alphas]
+    
+    # Iterate through alphas and compute wildcard feasibilities
+    feasible = []
+    for alpha in alphas:
+        wcm = _ddist_wildcard_model(alpha, ddists)
```

Comment:
Just a note: we could avoid creating a new budget object on each iteration here if we wanted/needed the performance.  I think this is fine for now - just an opportunity for later on if performance becomes an issue.